lib/view_component/css_module.rb

```diff
@@ -0,0 +1,119 @@
+# frozen_string_literal: true
+
+require "sass"
+
+module ViewComponent
+  class CSSModule < Sass::Tree::Visitors::Base
+    def initialize(module_name, raw_css)
+      @module_name, @mappings = module_name, {}
+
+      @css_root = Sass::SCSS::CssParser.new(raw_css, "(CSSModules)", 1).parse
+
+      @module_hash = compute_hash
+    end
+
+    # Given a `module_name` and `raw_css`, rewrite `raw_css`
+    # apply opaque transformations to `raw_css` so that
+    # selectors can only be accessed programatically,
+    # not by class name literals.
+    def self.rewrite(*args)
+      new(*args).rewrite
+    end
+
+    def rewrite
+      Sass::Tree::Visitors::SetOptions.visit(@css_root, {})
+
+      {
+        mappings: mappings,
+        css: @css_root.render
+      }
+    end
+
+    private
+
+    # Generate a short, random-ish token to prevent CSS selector collisions.
+    def compute_hash
+      Digest::MD5.hexdigest(@module_name).first(5) # rubocop:disable GitHub/InsecureHashAlgorithm
```

Comment:
What about using the first 5 characters of the Base64 representation of the hexdigest? This way there's more "specificity" in 5 chars, and even less risk of collision.

- inspired by `css-loader`'s `localIdentName: "[path][name]__[local]--[hash:base64:5]"` ([source](https://github.com/webpack-contrib/css-loader/blob/f1e0f00472fdd8fc1971c0602665b357b339e3b8/README.md#object))
- see https://stackoverflow.com/a/9987466/1789900 for an example of implementation